### PR TITLE
Feat/core

### PR DESCRIPTION
**Pull request by Jemoka**
# Pull Request

## Description
Core functionality of the astra package

### Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [x] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation update
- [ ] 🔧 Refactoring (no functional changes)
- [ ] ⚡ Performance improvement
- [ ] 🧪 Test additions or improvements
- [ ] 🏗️ Infrastructure/build changes

## Changes Made

- implements astra core data structures: environment, moderator, algorithm, problem etc.
- implements `DPO` algorithm, and `AST` environment (basic AST implementation)
- implements training harness
- implements `Detoxify` moderator
- implements a Huggingface extension adaptor that removes a lot of boilerplate for HF models
- writes 2 examples with GPT2 vs GPT2 vs Detoxify

## Sharp Edges
The extension module `astra_rl.ext` is not type checked, since many of the extension packages (for instance Huggingface's transformers) [does not yet](https://discuss.huggingface.co/t/static-type-checking-with-mypy-whats-the-official-position/464/4) support type checking.

### CODE REVIEW

File: examples/ast.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+problem = GPT2VsGPT2vsDetoxify("cuda")
+env = ASTEnvironment(problem, PROMPTS)
+
+# instantiate our solution
+solver = DPO(problem)
+optimizer = AdamW(problem.parameters(), lr=1e-5)
+
+# this is a training harness, from which we can call various functions to
+# handle training details
+harness = Harness(env, solver, batch_size=4, num_episodes_per_experience=2)
```
Comment (duncaneddy) [inline examples/ast.py:139]: Can you explain about the `Harness` a bit more? From discussions I was expecting a `Trainer` with a `.train()` method.

Is your thinking that `Harness` is more flexible in terms of allowing people to implement things themselves and even forces them to do so, while `Trainer.train()` is opinionated and makes decisions on behalf of users?

I don't think this should preclude the implementation of a `Trainer` so I think makes sense to have a lower level entrypoint.

Small nit maybe the name should be `TrainingHarness`?

File: src/astra_rl/ext/transformers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from .hf_ast_problem import HFASTProblem
```
Comment (duncaneddy) [inline src/astra_rl/ext/transformers/__init__.py:1]: What's the thinking behind the hierarchy of `ext/transformers/`?

Would something like `astra_rl/ext/huggingface.py` or `astra_rl/problems/huggingface.py` be more intuitive?

`transformers` seems to be a very-package-specific/implementation grouping that seems a little misaligned for storing HFProblem to me.

File: src/astra_rl/core/environment.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+from abc import abstractmethod, ABC
+from dataclasses import dataclass
+from typing import Sequence, Generic, Self, Optional
+
+from astra_rl.core.common import StateT, ActionT
+from astra_rl.core.problem import Problem
+
+
+@dataclass
+class Node(Generic[StateT, ActionT]):
```
Comment (duncaneddy) [inline src/astra_rl/core/environment.py:15]: Would be useful to do a scrub of the PR for everything that's a definition to add docstrings for the class and method definitions to capture the context and information from the implementation and reasoning while it's all still fresh.

File: src/astra_rl/core/common.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+"""
+common.py
+Common data structures.
+"""
+
+from typing import TypeVar
+
+StateT = TypeVar("StateT")
```
Comment (duncaneddy) [inline src/astra_rl/core/common.py:8]: See comment below, but docstrings and/or comments would help here to understand what these are.

File: src/astra_rl/core/algorithm.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+from typing import Sequence, Generic
+
+import torch
+
+from astra_rl.core.problem import Problem
+from astra_rl.core.environment import Graph
+from astra_rl.core.common import Step, Batch, StateT, ActionT
+
+
+class Algorithm(ABC, Generic[StateT, ActionT, Step, Batch]):
```
Comment (duncaneddy) [inline src/astra_rl/core/algorithm.py:15]: docstrings

File: src/astra_rl/algorithms/dpo.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+@dataclass
+class DPOBatch(Generic[StateT, ActionT]):
+    prefixes: Sequence[StateT]
+
+    suffix_pos: Sequence[ActionT]
+    suffix_neg: Sequence[ActionT]
+
+
+class DPO(
```
Comment (duncaneddy) [inline src/astra_rl/algorithms/dpo.py:29]: Docstrings

File: src/astra_rl/algorithms/dpo.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+@dataclass
+class DPOStep(Generic[StateT, ActionT]):
+    prefix: StateT
+
+    suffix_pos: ActionT
+    suffix_neg: ActionT
+
+
+@dataclass
+class DPOBatch(Generic[StateT, ActionT]):
```
Comment (duncaneddy) [inline src/astra_rl/algorithms/dpo.py:22]: Would be useful to do a scrub of the PR for everything that's a definition to add docstrings for the class and method definitions to capture the context and information from the implementation and reasoning while it's all still fresh.

**Review by duncaneddy — APPROVED** [review]
Looks great!

Left a couple comments for just my own personal clarification.

The only issue I see is there are a bunch of things class defs and methods that could still use detailed docstrings that reflect the reasoning and motivations of the pairing / design sessions so that we capture it while it is still fresh.

Up to you on whether you'd like to address / stub out here or just have us all do as a fast follow.

Feel free to mash merge whenever you're ready.

File: examples/ast.py (hunk 1 above)
Comment (Jemoka) [reply examples/ast.py:139]: Fair! Let's discuss a bit offline about how we want to do this. `trainer.train()` is a great API, but this means making optimization decisions for the user (optimizer, parameters, etc.) 
One thing I thought of was, perhaps, our trainer can do some kind of LR warmup; however, that seems like very custom as a function of the problem. Supporting control for something like this (+others) may require a whole lot of design decisions + parameters.

Happy to implement, but we should be careful what the degrees of control we offer to the user. 

`Harness` here, on the other hand, just offers high level APIs into environment and solver; that is, it gives a uniform API that the users can call to get a dataloader and get gradients though. It doesn't make any optimization decisions, and hence it really isn't much of a trainer.

File: src/astra_rl/ext/transformers/__init__.py (hunk 2 above)
Comment (Jemoka) [reply src/astra_rl/ext/transformers/__init__.py:1]: This is very fair! However, transformers does [not yet](https://discuss.huggingface.co/t/static-type-checking-with-mypy-whats-the-official-position/464/4) support typed Python, so I wanted to isolate the unpureness to a separate folder. This is noted in the "sharp edges" above.

**Comment by Jemoka** [comment]
Done, added documentation; lmk if this looks good + we can ship. Thanks!
